### PR TITLE
[Opaque Pointer] Revert some reverts and API usage removal

### PR DESCRIPTION
**Pull request by jsji**
- Revert "Revert "[SROA] Break typed pointer support""
- Revert "Revert "[LTO] Remove OpaquePointers option from config (NFC)""
- Revert "Revert "[LLVMContext] Remove hasSetOpaquePointersValue() API (NFC)""
- Revert "Make opaque pointer mode the default for 'llvm-c-test --echo'."
- Revert "Fix commit 4f7e5d220 for typed pointers"
- API usage removal

### CODE REVIEW

**Comment by elizabethandrews** [comment]
Is this part of enabling opaque pointers? What are these commits and why were they reverted in the first place?

**Comment by jsji** [comment]
> Is this part of enabling opaque pointers? What are these commits and why were they reverted in the first place?

The is part of `opaque pointer` *clean up*. These commits were added during transition as we were not ready for opaque pointer yet then, now we are all opaque pointers now, so should revert the revert to sync with upstream.

**Comment by jsji** [comment]
For the reverts, I think maybe the authors of revert can help confirm that they are no longer needed and can be reverted now. 
Authors, can you please help to review and confirm. Thanks.

- [ ] [Revert "Revert "[SROA] Break typed pointer support""](https://github.com/intel/llvm/pull/11175/commits/dc8263349c0db7ec456dcb3bda0d445191aa3ab1) @jcranmer-intel 
- [ ] [Revert "Revert "[LTO] Remove OpaquePointers option from config (NFC)""](https://github.com/intel/llvm/pull/11175/commits/a02819a25bb45685c7de35c753d917bd634e160e) @KornevNikita 
- [ ] [Revert "Revert "[LLVMContext] Remove hasSetOpaquePointersValue() API …](https://github.com/intel/llvm/pull/11175/commits/d543cc91f8c755f14991bac8054bc79d3db42974) @KornevNikita 
- [ ] [Revert "Make opaque pointer mode the default for 'llvm-c-test --echo'."](https://github.com/intel/llvm/pull/11175/commits/16a66eb2be100566c42143cc0c2c276db35c9c42) @cchen15 
- [ ] [Revert "Fix commit](https://github.com/intel/llvm/pull/11175/commits/1a81c4cb09d08c4d08a22200ce5c42747865fe0e) https://github.com/jsji/llvm/commit/4f7e5d22060e8a89237ffb93c3e7be6e92fee8fe [for typed pointers"](https://github.com/intel/llvm/pull/11175/commits/1a81c4cb09d08c4d08a22200ce5c42747865fe0e) @clin111

**Review by cchen15 — COMMENTED** [review]
LGTM. Thanks.

**Comment by elizabethandrews** [comment]
> > Is this part of enabling opaque pointers? What are these commits and why were they reverted in the first place?
> 
> The is part of `opaque pointer` _clean up_. These commits were added during transition as we were not ready for opaque pointer yet then, now we are all opaque pointers , so should revert the revert to sync with upstream.

Oh I see. Thanks for explaining.

File: llvm/lib/SYCLLowerIR/ESIMD/LowerESIMDVLoadVStore.cpp [hunk 1]
```diff
@@ -102,17 +102,15 @@ PreservedAnalyses ESIMDLowerLoadStorePass::run(Function &F,
       if (AS0 != AS1) {
         IRBuilder<> Builder(&Inst);
         if (GenXIntrinsic::isVStore(&Inst)) {
-          auto PtrTy = cast<PointerType>(Inst.getOperand(1)->getType());
-          PtrTy = PointerType::getWithSamePointeeType(PtrTy, AS1);
+          auto PtrTy = PointerType::get(Inst.getOperand(1)->getContext(), AS1);
```
Comment (asudarsa) [inline llvm/lib/SYCLLowerIR/ESIMD/LowerESIMDVLoadVStore.cpp:105]: From the original code of getWithSamePointeeType, it seems like the new code should be 
auto PtrTy = PointerType::get(Inst.getOperand(1)->getType()->getContext(), AS1);
Can you please check?

Thanks

**Review by asudarsa — COMMENTED** [review]
SYCLLowerIR changes look ok. Please take a look at one of my comments.

Thanks

**Review by jcranmer-intel — COMMENTED** [review]
SROA reversion LGTM.

**Review by clin111 — APPROVED** [review]
Thanks for catching it, it's good to remove this code now.
In fact, I found that it got merged to xmain, there is actually a double-fix in the xmain version.

File: llvm/lib/SYCLLowerIR/ESIMD/LowerESIMDVLoadVStore.cpp (hunk 1 above)
Comment (jsji) [reply llvm/lib/SYCLLowerIR/ESIMD/LowerESIMDVLoadVStore.cpp:105]: I think they should be the same, but yeah, updated to getType()->getContext() for safe.

**Comment by KornevNikita** [comment]
> For the reverts, I think maybe the authors of revert can help confirm that they are no longer needed and can be reverted now. Authors, can you please help to review and confirm. Thanks.
> 
> * [x]  [Revert "Revert "[SROA] Break typed pointer support""](https://github.com/intel/llvm/pull/11175/commits/dc8263349c0db7ec456dcb3bda0d445191aa3ab1) @jcranmer-intel
> * [ ]  [Revert "Revert "[LTO] Remove OpaquePointers option from config (NFC)""](https://github.com/intel/llvm/pull/11175/commits/a02819a25bb45685c7de35c753d917bd634e160e) @KornevNikita
> * [ ]  [Revert "Revert "[LLVMContext] Remove hasSetOpaquePointersValue() API …](https://github.com/intel/llvm/pull/11175/commits/d543cc91f8c755f14991bac8054bc79d3db42974) @KornevNikita
> * [x]  [Revert "Make opaque pointer mode the default for 'llvm-c-test --echo'."](https://github.com/intel/llvm/pull/11175/commits/16a66eb2be100566c42143cc0c2c276db35c9c42) @cchen15
> * [x]  [Revert "Fix commit](https://github.com/intel/llvm/pull/11175/commits/1a81c4cb09d08c4d08a22200ce5c42747865fe0e) [jsji@4f7e5d2](https://github.com/jsji/llvm/commit/4f7e5d22060e8a89237ffb93c3e7be6e92fee8fe) [for typed pointers"](https://github.com/intel/llvm/pull/11175/commits/1a81c4cb09d08c4d08a22200ce5c42747865fe0e) @clin111

I'm okay with changes made by me

**Review by KornevNikita — APPROVED** [review]

**Review by smanna12 — APPROVED** [review]

**Comment by jsji** [comment]
@intel/llvm-gatekeepers Can we get this merged. Thanks!

**Pull request merged**